### PR TITLE
Fix deploying jobs on 1.6 kubernetes

### LINKED ISSUE #94: Test Kit-deployer status checking with Kuberetes Server version 1.6.x
This requires #92 first. We will need to test the current deployment status checks work with the 1.6 deployments notification. This should be seamless but we need to validate it. Requires a working Cluster with version 1.6.x

### LINKED ISSUE #109: Failure deploying job on 1.6 cluster
When deploying a `job` to a kubernetes 1.6 cluster we are seeing the following error message. However the job is still successfully deployed. So the assumption is that this is thrown from the health check status checking and this is done using web API requests.

```
[ERROR] LogID: ed8d0c10-16ff-4c97-8739-9815054495cb Task: deploy - kit-deployer - Error: the server could not find the requested resource
```

### PR DESCRIPTION
**Pull request by chesleybrown**
# What

- fixes #109 and fixes #94 
- we need to use the `batch` api for deployments and not the `extensions` api
- switched functional tests to use `minikube` which is more stable
- functional tests are now faster by allowing you to run the tests without completely destroying and re-creating the kubernetes cluster
- I downgraded the kubectl from `1.6.7` to `1.6.4` as that was the latest `1.6` version supported by minikube and I wanted to make sure it was working as we expected

# Testing

- Install [minikube](https://kubernetes.io/docs/tasks/tools/install-minikube/)
- Run `./test-functional`

### CODE REVIEW

File: src/lib/kubectl.js [hunk 1]
```diff
@@ -113,54 +113,73 @@ class KubectlEventWatcher extends EventEmitter {
   }
 }
 
-class Kubeapi {
-  constructor(cwd, kubeconfig) {
+class Current {
```
Comment (c1freitas) [inline src/lib/kubectl.js]: `Currect`? You have been working with Rosie to long :P

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -30,7 +30,8 @@ Alternatively you can use docker-compose:
 
 We were unable to get the functional tests working via Codeship, so instead you will have to test via the `test-functional` command (which uses `docker-compose` internally). This will spin up a Kubernetes cluster and then run the functional tests against it and then shutdown and delete the cluster afterwards.
 
-1. `./test-functional`
+1. First install [minikube](https://kubernetes.io/docs/tasks/tools/install-minikube/)
```
Comment (manute) [inline CONTRIBUTING.md]: this is nice. We might do these tests for all our projects that operates with kubernetes

File: src/lib/kubectl.js [hunk 3]
```diff
@@ -113,54 +113,73 @@ class KubectlEventWatcher extends EventEmitter {
...
+    this.currentUser = _.find(kubeconfig.users, {
+      name: this.currentContext.context.user
     });
-    if (!currentUser) {
+    if (!this.currentUser) {
       throw new Error("Unable to configure kubeapi with current user");
     }
+  }
+  get context() {
+    return this.currentContext;
```
Comment (manute) [inline src/lib/kubectl.js]: oh welcome to the encapsulation problem

File: src/lib/kubectl.js [hunk 4]
```diff
@@ -261,39 +281,67 @@ class Kubectl extends EventEmitter {
...
+                  if (name) {
+                    cmd.push(name);
+                  }
+                  this.spawn(cmd, (spawnErr, spawnData) => {
+                    if (spawnErr) {
+                      return reject(new Error(spawnErr));
+                    }
+                    return resolve(JSON.parse(spawnData));
+                  });
+                } else {
```
Comment (manute) [inline src/lib/kubectl.js]: sorry I have lost with so many `if/else/try/catch` 😅

File: src/lib/kubectl.js (hunk 1 above)
Comment (chesleybrown) [reply src/lib/kubectl.js]: giphy

File: src/lib/kubectl.js (hunk 4 above)
Comment (chesleybrown) [reply src/lib/kubectl.js]: I know. It's horrible. 🙈

Comment (c1freitas) [reply src/lib/kubectl.js]: That promise chain makes my head hurt 🤕

Comment (chesleybrown) [reply src/lib/kubectl.js]: Sorry. I'll look into refactoring if I have to revisit this again (which I hope I don't...).

**Comment by brianhall** [comment]
Nice work, @chesleybrown. Once this is deployed and confirmed to be working, can you let the data team and CCB know that it's good to go? I believe they're the ones most aware that this was broken.

**Pull request merged**